File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/codec/EnumCodec.java
package net.volcano.jdacommands.model.command.arguments.implementation.codec;

import net.volcano.jdacommands.model.ParameterData;
import net.volcano.jdacommands.model.command.arguments.implementation.argument.EnumArgument;
import net.volcano.jdacommands.model.command.arguments.interfaces.Codec;
import net.volcano.jdautilities.utils.ListUtilKt;
import net.volcano.jdautilities.utils.StringUtilKt;
import org.springframework.stereotype.Component;

import java.util.Arrays;

@Component
public class EnumCodec extends Codec<Enum<?>> {
	
	@Override
	public EnumArgument buildArgument(ParameterData data) {
		var builder = EnumArgument.builder();
		
		if (!(data.parameter.getParameterizedType() instanceof Class)) {
			throw new IllegalArgumentException("Parameter type not instance of class");
		}
		
		var enumClass = ((Class<Enum>) data.parameter.getParameterizedType());
		
		if (enumClass.isArray()) {
			enumClass = (Class<Enum>) enumClass.componentType();
		}
		
		var constants = enumClass.getEnumConstants();
		
		if (constants.length == 0) {
			throw new IllegalArgumentException("Cannot have empty enum as argument");
		}
		
		builder.options(enumClass.getEnumConstants());
		
		builder.usage(ListUtilKt.asString(Arrays.asList(constants), " | ", e -> StringUtilKt.capitalize(e.name().toLowerCase())));
		
		return builder.build();
	}
	
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/codec/EnumCodec.kt
package net.volcano.jdacommands.model.command.arguments.implementation.codec

import net.volcano.jdacommands.model.ParameterData
import net.volcano.jdacommands.model.command.arguments.implementation.argument.EnumArgument
import net.volcano.jdacommands.model.command.arguments.interfaces.Codec
import net.volcano.jdautils.utils.capitalize
import net.volcano.jdautils.utils.enumConstants
import net.volcano.jdautils.utils.isEnum
import org.springframework.stereotype.Component

@Component
class EnumCodec : Codec<Enum<*>?>() {

	public override fun buildArgument(data: ParameterData): EnumArgument {

		require(data.parameter.type.isEnum) { "Parameter type is not enum." }

		val constants = data.parameter.type.enumConstants

		require(constants != null) { "Parameter type is not enum." }
		require(constants.isNotEmpty()) { "Cannot have empty enum as argument." }

		val arg = EnumArgument(constants as Array<Enum<*>>)
		arg.usage = constants.joinToString(" | ") { it.name.lowercase().capitalize() }

		return arg
	}
}